# bnigra.pieris.PEK

This dataset is part of the preprint "A butterfly egg-killing hypersensitive response in Brassica nigra is controlled by a single locus, PEK, containing a cluster of TIR-NBS-LRR receptors"
doi: https://doi.org/10.22541/au.168942068.87319933/v1 

Title of the dataset:
<br>[Data from:] A butterfly egg-killing hypersensitive response in Brassica nigra is controlled by a single dominant PEK locus (Pieris egg-killing) containing a cluster of TIR-NBS-LRR receptors
<br>doi: https://doi.org/10.5281/zenodo.8131352

Creator: 
<br>Name:	Niccol√≤ Bassetti
<br>Institution: Biosystematics Group (Wageningen University & Research)
<br>ORCID: https://orcid.org/0000-0003-1859-4698
<br>Email: niccolo.bassetti@protonmail.com

Abstract:
<br>In this study, we found that P. brassicae egg-induced HR segregates as a Mendelian trait in wild accessions of black mustard B. nigra L. Through bulk-segregant analysis coupled with whole-genome sequencing (BSAseq),
we identified a single dominant locus on chromosome B3 which we named PEK (Pieris egg-killing locus).

Data and file overview:
<br>Files are arranged in the following folders, with each folder referrring to a figure of the manuscript/article.
<br>Tables are only presented in the manuscript and in the Zenodo repository.

figure_1:
<br>Figure1_2023.R - R script for statistical analysis
<br>RawdataFigure1.csv - dataset used for R script "Figure1_2023.R"
<br>Figure1.xlsx - excel file used to plot final figure
	
figure_2:
<br>fig.2a_DG1-S1_vs_SF48-O1.jpg - raw image used for panel 1a
<br>fig.2b_README.txt - README explaining that raw data for figure 2b are included in folder table_1_and_S1
<br>fig.2c_dataset_BSA_results.xlsx - raw data from BSA-seq analysis and plotting
<br>fig.2c_dataset_markers_BC1.xlsx -  - raw marker data plotted in figure 2C
<br>fig.2c_script_BSA-seq.bash - BASH script used for analysis of BSA-seq data
<br>KASP-01_marker trait association.R - R script for statistical analysis
<br>rqtl_HR_BC1.csv - dataset used for R script "KASP-01_marker trait association.R"
			
figure_3:
<br>fig.3a.xlsx - raw marker data plotted in figure 3A
<br>fig.3b.xlsx - raw marker data for 64 BC1-S1 recombinants used for fine-mapping 
<br>fig.3b_README.txt - README explaining how marker data were assembled to generate figure 3b
<br>KASP-02_marker trait association.R - - R script for statistical analysis
<br>rqtl_HR_BC1S1_full_population.csv - dataset used for R script "KASP-02_marker trait association.R"

figure_4:
<br>fig.4.png - figure representing a model proposed by the article

figure_S1:
<br>fig. S1.pptx - Powerpoint with images/drawings

figure_S2:
<br>figure_S2.xlsx - excel file used to plot final figure
<br>figure_S2_data_analysis.R - R script for statistical analysis
<br>figure_S2_dataset.txt - dataset used for R script "figure_S2_data_analysis.R"
			
figure_S3:
<br>fig.S3.png - figure reported in the article

figure_S4:
<br>fig.S4_README.txt - README file explaining that data for this figure are already located in folder "figure_2"

figure_S5:
<br>figure_S5.png - figure reported in the article
<br>figure_S5_README.txt - README file explaining how to generate figure_S5.png
	
figure_S6:
<br>figure_S6_README.txt - README file explaining how to use Excel files within folder "figure_S6"
<br>figure_S6_full_population.xlsx - raw marker data and statistical analysis of phenotypic data for populations BC1S2
<br>figure_S6_recombinants.xlsx - raw marker data for recombinants of population BC1S2

figure_S:
<br>figure_S7_README.txt - README file explaining Synmap analysis in COGE
<br>fig.S7a.png	- results from Synmap analysis in COGE 
<br>fig.S7b.png	- results from Synmap analysis in COGE 


This dataset is published under the CC BY (Attribution) license.
<br>This license allows reusers to distribute, remix, adapt, and build upon the material in any medium or format, so long as attribution is given to the creator.



